demos/unity_states/unity/Assets/Scripts/BallStateBillboardViewer.cs

```diff
@@ -0,0 +1,29 @@
+using UnityEngine;
+using UnityEngine.UI;
+
+public class BallStateBillboardViewer : MonoBehaviour
+{
+    [SerializeField] private bool isPlatform;
+    [SerializeField] private BallStateDataSourcePlatform dataSourcePlatform;
+    [SerializeField] private BallStateDataSourceUnity dataSourceUnity;
+    [SerializeField] private Text typeView;
+    [SerializeField] private Text detailsView;
+
+    private IBallStateDataSource dataSource;
+
+    void Awake()
+    {
+        this.typeView.text = this.isPlatform ? "PLATFORM" : "UNITY";
+        this.dataSource = this.isPlatform ? this.dataSourcePlatform : this.dataSourceUnity;
```

Comment:
since we do this, does it still make sense to have those other fields? Why not just get the component here and require that those components exist?

I'm not familiar with which pattern would be better in unity so pls advise

Reply:
Yes; those scripts are attached to other GameObjects in the scene. For the type and details views, it is also possible to get the component like this: `this.typeView = this.transform.GetChild(0).GetComponent<Text>()`.

However, that hardcodes a hierarchical relationship between objects, is less ergonomic, and is more prone to NREs.

![Screen Shot 2022-08-05 at 4 39 37 PM](https://user-images.githubusercontent.com/5103968/183223957-8591828d-b4d3-4ce1-b96a-b79d3757d813.png)

Reply:
I didn't understand - I was asking about getting GetComponent to get the dataSourcePlatform or dataSourceUnity depending on this.isPlatform, instead of storing those as fields

Reply:
ah ok, I think I understand - the components don't exist on the billboard, so thats what you meant by heirarchy - we have to traverse the gameobject heirarchy to get them. If we just store the fields, we can assign them in the editor instead. 

Mark as resolved when read.

Reply:
Yep